### PR TITLE
JAVA-3839

### PR DESCRIPTION
**Pull request by jyemin**
@rozza this is preliminary work as the Scala build fails when running against Java 14.  See https://jira.mongodb.org/browse/JAVA-3839 for the exception I get.  I found https://github.com/sbt/sbt/issues/5093 but I'm not sure what, if anything, can be updated in our build script to take advantage of the fix.  Can you take a look?

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -311,22 +311,24 @@ configure(javaCodeCheckedProjects) {
...
     }
 
     codenarc {
-        toolVersion = '1.1'
+        toolVersion = '1.6.1'
         reportFormat = project.buildingWith('xmlReports.enabled') ? 'xml' : 'html'
     }
 
-    tasks.withType(com.github.spotbugs.SpotBugsTask) {
+    spotbugsMain {
```
Comment (jyemin) [inline build.gradle:321]: It complained without this change

File: build.gradle [hunk 2]
```diff
@@ -311,22 +311,24 @@ configure(javaCodeCheckedProjects) {
...
         reports {
             xml.enabled = project.buildingWith('xmlReports.enabled')
             html.enabled = !project.buildingWith('xmlReports.enabled')
         }
     }
 
+    spotbugsTest {
```
Comment (jyemin) [inline build.gradle:328]: It complained about the use of sourceSets up above so using this as an alternative.

File: config/codenarc/codenarc.xml [hunk 3]
```diff
@@ -55,6 +55,13 @@
         <exclude name='MethodParameterTypeRequired'/>
         <exclude name='FieldTypeRequired '/>
         <exclude name="VariableTypeRequired"/>
+        <exclude name="CompileStatic"/>
```
Comment (jyemin) [inline config/codenarc/codenarc.xml:58]: Excluding any new rules added since 1.1

File: build.gradle [hunk 4]
```diff
@@ -26,7 +26,7 @@ buildscript {
     }
     dependencies {
         classpath 'com.netflix.nebula:gradle-extra-configurations-plugin:3.0.3'
-        classpath "com.github.spotbugs:spotbugs-gradle-plugin:3.0.0"
+        classpath "gradle.plugin.com.github.spotbugs.snom:spotbugs-gradle-plugin:4.5.0"
```
Comment (jyemin) [inline build.gradle:29]: Not sure why the name change, but that's what it is now.

Comment (rozza) [reply build.gradle:29]: Ack

File: build.gradle (hunk 1 above)
Comment (rozza) [reply build.gradle:321]: Ack

File: build.gradle (hunk 2 above)
Comment (rozza) [reply build.gradle:328]: Ack

File: config/codenarc/codenarc.xml (hunk 3 above)
Comment (rozza) [reply config/codenarc/codenarc.xml:58]: Ack

**Review by rozza — CHANGES_REQUESTED** [review]
Update `gradle.properties`:

```
scalaVersions=2.11.12,2.12.12,2.13.3
defaultScalaVersions=2.13.3
```

See: https://docs.scala-lang.org/overviews/jdk-compatibility/overview.html#version-compatibility-table

You'll have to edit the publishing the scala snapshots / archives and set the jdk to 11 for the 2.11 & 2.12 versions of scala.

**Review by rozza — COMMENTED** [review]
Typo: Should be `2.13.3`

**Review by rozza — APPROVED** [review]
LGTM!

**Comment by jyemin** [comment]
Added one more update to fix an issue when using Java 15.

**Review by rozza — APPROVED** [review]

**Pull request merged**